### ISSUE TITLE
Simultaneous code loading from multiple Julia processes is unsafe

### ISSUE THREAD

**Issue by LilithHafner** [JuliaLang/julia#52265]
I ran into an issue when [running CI for a package](https://github.com/LilithHafner/RegressionTests.jl/actions/runs/6950232743/job/18909981510?pr=5#step:6:107) where I often, but not always, get the following error

```
ERROR: Error opening package file /home/runner/.julia/compiled/v1.9/RegressionTests/L2wCO_j3iXv.so: /home/runner/.julia/compiled/v1.9/RegressionTests/L2wCO_j3iXv.so: cannot open shared object file: No such file or directory

Stacktrace:
 [1] _include_from_serialized(pkg::Base.PkgId, path::String, ocachepath::String, depmods::Vector{Any})
   @ Base ./loading.jl:1052
 [2] _tryrequire_from_serialized(pkg::Base.PkgId, path::String, ocachepath::String)
   @ Base ./loading.jl:1442
 [3] _require(pkg::Base.PkgId, env::String)
   @ Base ./loading.jl:1816
 [4] _require_prelocked(uuidkey::Base.PkgId, env::String)
   @ Base ./loading.jl:1660
 [5] macro expansion
   @ ./loading.jl:1648 [inlined]
 [6] macro expansion
   @ ./lock.jl:267 [inlined]
 [7] require(into::Module, mod::Symbol)
   @ Base ./loading.jl:1611
 [8] top-level scope
   @ none:1
```

The package is launching several separate, uncoordinated Julia processes which load multiple versions of the same package. From looking at the stack trace, I think the issue is that code loading uses locks to coordinate access to shared object files but that locking is not at the level of the data being accessed so it is possible for multiple Julia processes to take locks to that data simultaneously. (I'm also seeing hangs which may or may not be related)

**Comment by vchuravy** [comment]
Fixed by #49052 afaik

**Comment by LilithHafner** [comment]
I set `--compiled-modules=no` for versions of Julia less than 1.10 and have not seen that error since.
